Initial version of class register

Agata-Gizinska opened the pull request:
The aim of this program is to handle data about students, classrooms and courses.
The data are stored in 3 csv-based registers:
- ClassroomRegister - contains classroom name, start and end years, a list of active students, a list of graduates and a list of drop-outs.
- StudentRegister - contains student's data, such as name, date of birth, assigned classroom and assigned courses.
- CourseRegister - cointains course name, a required number of grades for a student to finish the course, a list of active students, a list of graduates of the course and a list of drop-outs of the course.

The program can process data and print them into the console. Moreover, the user should be able to add a new student, add new classroom, add new courses, give grades to students for courses that they attend.
The initial version implement basic classes used by the program: 
- Classroom, Student, Course - used to manipulate data inside the program
- Register (as a parent class) and child classes: ClassroomRegister, StudentRegister, CourseRegister - used to manipulate data which will be stored in csv files.

The program should be accessible to use from command prompt, so the initial version contains argument parser stored in main() function. For now, it is possible to print all registers (print_class, print_students, print_courses) and add a new student to all registers (new_student).

Agata-Gizinska committed:
Initial version of class_register; added class Classroom [implemented __init__, add_student, drop_out_student], added class Student [implemented __init__, __repr__, add_course, drop_out], class Course [implemented __init__, __repr__, give_grade, pass_course]

Agata-Gizinska committed:
Added: Classroom.__repr__; Implemented: Classroom._graduate_student(), Student.get_grade(), Student.calc_final_grade(), Student.graduate(), Course.give_grade(), Course.pass_course()

Agata-Gizinska committed:
Added class Register and its child classes ClassroomRegister, StudentRegister, CourseRegister to handle manipulating new csv databases (classrooms.csv, students.csv, courses.csv). Added add_student_to_register methods in classes Classroom, Student, Course. Added main() function which enables parsing arguments in a command prompt.

Agata-Gizinska committed:
Implementation of “add new student” feature: Implemented add_student_to_classroom method in ClassroomRegister class; Implemented new_student method in StudentRegister class; Implemented add_student_to_course method in CourseRegister class; Implemented parser new_student argument handler

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                if classroom.name in row:
+                    row_index = rows.index(row)
+                else:
+                    pass
```
```
for row in rows:
  if classroom.name in row:
    row_index = rows.index(row)
    break # break exits the loop early so we stop iteration when row_index is found
````

remove else clause it does not do anything

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                # Overwrite Classroom Register with updated list of entries
+                writer.writerows(rows)
+            else:
+                pass
```
Redundant, just not entering the if clause will have the same effect

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                rows.insert(row_index, new_row)
+                # Reset whole Classroom Register
+                file.seek(0)
+                file.truncate(0)
```
'r+' does not allow trucating file open with 'rw+'
Also opening with write enabled will truncate file so you can remove this line

Agata-Gizinska replied:
File "C:/Users/User/class_register.py", line 39, in add_student_to_classroom
    with open(self.file, 'rw+', newline='', encoding='utf-8') as file:
ValueError: must have exactly one of create/read/write/append mode

Python 3 does not seem to be supporting 'rw+' mode... you can either 'r+' or 'w+', but the problem is that 'w+' truncates whole csv file BEFORE you get data from it - that doesn't do good, because I need to get data from csv first, then process the data that I can append new student info and THEN truncate file and write it with updated data. That is why csv file is opened in 'r+' and truncate is explicitly called.

czarnadzuma11 replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+
+    def new_student(self, *args):
+        # create new Student instance
+        new_student_ = Student(*args)
```
_new_student

Agata-Gizinska replied:
why should I make it a private variable? The trailing underscore was meant to make the variable differ from function name. If it's confusing I can change the variable name

czarnadzuma11 replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+        pass
+
+    def add_student_to_course(self, student, course):
+        with open(self.file, 'r+', newline='', encoding='utf-8') as file:
```
again, 'rw+' because you are truncating file

Agata-Gizinska replied:
again commented up

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                rows.insert(row_index, new_row)
+                # Reset whole Course Register
+                file.seek(0)
+                file.truncate(0)
```
Remove if you open with 'rw+'

Agata-Gizinska replied:
same

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+    def add_course(self, course):
+        self.courses.append({course: []})
+        course.attending_students.append(self)
+        return self.courses
```
Is there a point to returning self.courses from this function?

Agata-Gizinska replied:
removed

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                    pass
+                return self.courses
+            else:
+                pass
```
Again, 
```
else:
  pass
```
is equivalent to not adding else at all, applies to the enire pull request

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+    # add student to Course Register
+    @staticmethod
+    def add_student_to_register(student, register):
+        register.add_student_to_course(student, student.courses[0])
```
I do not see the point of wrapping  register.add_student_to_course(student, student.courses[0]) with add_student_to_register.
Just do register.add_student_to_course(student, student.courses[0]) and remove this function

Agata-Gizinska replied:
Ok, removed wrapper in both Course and Classroom

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+    parser.add_argument('--give_grade', '-grade', nargs=2,
+                        help="Give grade to student, give grade, "
+                             "student's name")
+
```
one dash parameters are called short and are supposed to be one letter long

Agata-Gizinska replied:
removed then

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+    elif args.new_student:
+        args_ = args.new_student
+        assigned_classroom = args_[3]
+        first_course = args_[4]
```
if elif else construction guarantees that only one of actions will be executed but your parser allows to specify multiple of them in one application call.

Either make parser accept only one action at a time https://docs.python.org/3/library/argparse.html#mutual-exclusion
or execute all specified actions (but then you will need to worry about the order in which they were passed)

czarnadzuma11 replied:
Do not need to fix if you move to subparsers

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+        first_course = args_[4]
+
+        # check if assigned classroom is in Classroom Register
+        def is_classroom_in_register(classroom):
```
this should be a ClassroomRegister method

Agata-Gizinska replied:
moved to ClassroomRegister

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                        students = row.get('Students').strip('[]').split(',')\
+                            if not "''" else []
+                        graduates = row.get('Graduates').strip('[]'). \
+                            split(',') if not "''" else []
```
```if not "''"```
does not compare to anything, you have to do
```if not ("''" == compared_to_value) else []``` or
```if "''" != compared_to_value else []```

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                        graduates = row.get('Graduates').strip('[]'). \
+                            split(',') if not "''" else []
+                        dropout = row.get('Dropout').strip('[]').split(',') \
+                            if not "''" else []
```
same

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                        course_name = row.get('Course name')
+                        grade_number = row.get('Grades to pass')
+                        students = row.get('Students').strip('[]').split(', ')\
+                            if not "''" else []
```
same

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                        students = row.get('Students').strip('[]').split(', ')\
+                            if not "''" else []
+                        graduates = row.get('Graduates').strip('[]'). \
+                            split(', ') if not "''" else []
```
same

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                        graduates = row.get('Graduates').strip('[]'). \
+                            split(', ') if not "''" else []
+                        dropout = row.get('Dropout').strip('[]').split(', ') \
+                            if not "''" else []
```
same

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+                        print('Create new classroom first!')
+
+        # check if first course is in Course Register
+        def is_course_in_register(course):
```
Again, why is this not course method

Agata-Gizinska replied:
moved to CourseRegister

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+        # append info about registers' objects to arguments, so program can
+        # use them later
+        args_.append(class_reg)
+        args_.append(course_reg)
```
don't do this, just pass them to new_student as named paramterers

Agata-Gizinska replied:
ok

czarnadzuma11 commented on class_register.py:
```diff
@@ -0,0 +1,438 @@
+    parser.add_argument('--new_classroom', '-newcl', nargs='*',
+                        help='Add new classroom, give start year, end year')
+    # add new student to Student Register
+    parser.add_argument('--new_student', '-newst', nargs=5,
```
Unacceptable, each of those fields should be a separate named argument.

Agata-Gizinska replied:
Could you please elaborate?

czarnadzuma11 replied:
Describing what and in what order has to be entered explains how to use it TO THE USER but does not explain int TO THE PROGRAM.
This is why in code below we need to guess that
args.new_student[0] means student_name

You might have noticed that each command here expects a different set of arguments, this is almost as if each command should have its own parser. And we can do that with subparsers!


```
import argparse
parser = argparse.ArgumentParser()
subparsers = parser.add_subparsers(dest='command')

subparser = subparsers.add_parser('new-student')
subparser.add_argument('name')
subparser.add_argument('surname')

subparser = subparsers.add_parser('delete-student')
subparser.add_argument('id')

subparser = subparsers.add_parser('list')

args = parser.parse_args()
print(args)

if args.command == 'list':
    list_students()
elif args.command == 'new-student':
    add_student(args.name, args.surname)
```

an alternative (lame but quick) solution woiuld be to manually put new_student values into for example a dictionary dict['name'] = args.new_student[0] etc but it would not verify arguments at parsing and would require you to handle input errors yourself

Agata-Gizinska replied:
ok

Agata-Gizinska committed:
Fix: changed outcome of else statement in is_classroom_in_register and is_course_in_register functions and set these functions that they exit program if prompted classroom or course don't exist; moved is_classroom_in_register function to ClassroomRegister; moved is_course_in_register function to CourseRegister; removed redundant else: pass statements; removed redundant add_student_to_register wrappers; removed single-dash parameters; fixed ternary statements in methods for ClassRegister and CourseRegister

Agata-Gizinska committed:
Introduced new feature Add new student; minor fixes in __init__ methods of Classroom and Course classes; added several descriptive comments

Agata-Gizinska committed:
Introduced Add new course feature; Fixed bug that prevented is_course_in_register method from finding existing course which is not in the first row of the register; minor unification of naming variables regarding the number of partial grades required to pass a course; minor change in output message if classroom or course don’t exist while creating new student

Agata-Gizinska committed:
Added read_rows_in_register method in Register and replaced ‘with open(file)’ statements in other class methods; added extract_classroom_info method in ClassroomRegister; minor changes and additions in placeholder methods; added is_student_in_register, is_student_attending_course and extract_student_info methods in StudentRegister; added extract_course_info in CourseRegister; changes in Student __init__ method: check if student object is already registered; minor formatting changes in ‘—new_student’ and ‘—new_course’ parser options; initial version of ‘—give_grade’ parser option

Agata-Gizinska committed:
docs: Changed some comments into docstrings and added some new docstrings throughout the program; Added some placeholder methods for dropping out student

Agata-Gizinska committed:
Fully functional '--give_grade' option

feat: Implemented update_student_info method of StudentRegister class;
Implemented move_student_to_graduates method of CourseRegister class;

fix: Changed csv.reader and csv.writer delimiter to semicolon to omit problems with lists in entries;
Updated extract_student_info method of StudentRegister class so it now handles data about courses properly;
Reformatted new_student method of StudentRegister class to csv.writer;
Updated Student class __init__ method so self.courses attribute is assigned properly despite the number of courses;
Updated get_grade method of Student class depending on the number of grades that student has already received for a course;
Updated ‘—new_student’ option to check if student is already in register;
Minor fixes in variables’ naming;
Minor fixes in string formatting throughout whole program;
Changed __repr__ methods in Classroom class;
Updated calc_final_grade method of Student class that strings are converted to integers;
Updated pass_course method of Course class;

docs: Minor changes in docstrings;
Added more descriptive comments.

Agata-Gizinska committed:
feat: enabled appending of an existing student to an existing course by adding ‘—append_to_course’ parser argument

fix:
- merged methods move_student_to_graduates and move_student_to_dropouts into change_student_status in ClassroomRegister class and in CourseRegister class;
- reformatted update_student_info method in StudentRegister class so it can be used when giving a grade to a student and adding a student to existing course;
- minor fix in add_student_to_course method of CourseRegister class;
- changed the sequence of methods in Classroom class;
- minor fix in __init__ method of Student class;
- updated add_to_course method in Student class;

Agata-Gizinska committed:
Bug fixes and started working on graduation feature

feat: began working on graduation feature:
- Implemented check_passed_courses method in Student class;
- Added new conditional statement in get_grade method in Student class;
fix:
- fixed bug in extract_student_info method in StudentRegister class;
- fixed bug in update_student_info method in StudentRegister class;
- fixed bugs in __init__ method in Student class;
- minor fix in add_to_course method in Student class;
- minor fix in add_student_to_course method in Course class.

Agata-Gizinska committed:
Graduation feature is operational

feat: implemented graduation feature;
fix:
- fixed sequencing of data in entries in change_student_status method in ClassroomRegister class;
- fixed bug in graduate_student method in Classroom class;
- fixed bug in graduate method in Student method;
- fixed bug in pass_course method in Course class;
docs:
- added informative print message that student has graduated

Agata-Gizinska committed:
Implemented drop out student feature

feat:
- implemented dropping out student feature
- implemented student's status in the Student Register
fix:
- grammar fixes in documentation
- improved naming of variables
- added student's status check to prevent giving grades to graduates and dropouts
docs:
- implemeted docstrings for classes
- improved docstrings for methods
- improved comments

Agata-Gizinska committed:
Fixes and reformatting

fix:
- reformatted long if-elif into functions,
- implemented subparsers for each main feature,
- removed redundant docstrings and comments

Agata-Gizinska committed:
Implemented subparsers into main function

fix:
- reformatted argparse main function by introducing subparsers for each feature;
- reformatted child Register classes that string formatting is handled in the parent Register class;
- changed csv.reader and csv.writer into csv.DictReader and csv.DictWriter when csv files are opened in reading or reading+writing mode, adjusted methods to handle dictionaries properly;
- renamed methods in ClassroomRegister to get_classroom_from_register;
- renamed methods in CourseRegister to get_course_from_register;
- added update_student_info method into graduate and drop_out methods in Student class;
- implemented a condition in append_to_course_func, preventing from adding the student to more courses than are necessary to graduate.
docs:
- adjusted docstrings to changes in classes;
- added docstrings to new functions;
- added help message to ‘print_register’ subparser

Agata-Gizinska committed:
Implemented subparsers into main function

fix:
- reformatted argparse main function by introducting subparsers for each feature;
- reformatted child Register classes that string formatting is handled in the parent Register class;
- changed csv.reader and csv.writer into csv.DictReader and csv.DictWriter when csv files are opened in reading or reading+writing mode, adjusted methods to handle dictionaries properly;
- renamed methods in ClassroomRegister to get_classroom_from_register;
- renamed methods in CourseRegister to get_course_from_register;
- added update_student_info method into graduate and drop_out medthods in Student class;
- implemented a condition in append_to_course_func, preventing from adding the student to more courses than are necessary to graduate.
docs:
- adjusted docstrings to changes in classes;
- added docstrings to new functions;
- added help message to ‘print_register’ subparser

Agata-Gizinska committed:
Merge remote-tracking branch 'origin' into devwork_new_features

Agata-Gizinska committed:
Merge pull request #3 from Agata-Gizinska/devwork_new_features

Improved version of the Class Register program
